File: 2019/python/day_02b.py
#!/usr/bin/python3

import copy
import sys


def main():
    # Get program
    file_name = "../input/day_02_input"
    desired_output = 19690720
    if len(sys.argv) > 1:
        file_name = sys.argv[1]
    file = open(file_name)
    noun_verb_list = list(range(0, 100))
    program_str = file.readline()

    # Solve
    ans = -1
    for noun in noun_verb_list:
        for verb in noun_verb_list:
            if noun == verb:
                continue
            program = list(map(int, program_str.split(",")))
            program[1] = noun
            program[2] = verb
            inst_ptr = 0
            while program[inst_ptr] != 99:
                if program[inst_ptr] == 1:
                    program[program[inst_ptr + 3]] = (
                        program[program[inst_ptr + 1]] + program[program[inst_ptr + 2]]
                    )
                elif program[inst_ptr] == 2:
                    program[program[inst_ptr + 3]] = (
                        program[program[inst_ptr + 1]] * program[program[inst_ptr + 2]]
                    )
                inst_ptr = inst_ptr + 4
            if program[0] == desired_output:
                ans = noun * 100 + verb
                print(ans)
                return ans
    return ans


if __name__ == "__main__":
    main()

File: 2019/cpp/day_02b.cpp
#include <fstream>
#include <iostream>
#include <iterator>
#include <numeric>
#include <string>
#include <vector>

int main(int argc, char *argv[]) {
  // Get input
  std::string input = "../input/day_02_input";
  if (argc > 1) {
    input = argv[1];
  }
  std::ifstream file(input);
  std::string input_str;
  const std::string delimiter = ",";
  std::getline(file, input_str);
  std::size_t start = 0;
  std::size_t end = input_str.find(delimiter);
  std::vector<int> program_main_copy;
  while (end != std::string::npos) {
    program_main_copy.emplace_back(
        std::stoi(input_str.substr(start, end - start)));
    start = end + delimiter.size();
    end = input_str.find(delimiter, start);
  }
  program_main_copy.emplace_back(
      std::stoi(input_str.substr(start, end - start)));

  long long desired_output = 19690720;
  std::vector<int> nv_list(100);
  std::iota(std::begin(nv_list), std::end(nv_list), 0);

  // Solve
  long long ans = -1;
  for (const auto noun : nv_list) {
    for (const auto verb : nv_list) {
      if (noun == verb) continue;
      auto program = program_main_copy;
      program[1] = noun;
      program[2] = verb;
      size_t index = 0;
      while (program[index] != 99) {
        if (program[index] == 1) {
          program[program[index + 3]] =
              program[program[index + 1]] + program[program[index + 2]];
        } else if (program[index] == 2) {
          program[program[index + 3]] =
              program[program[index + 1]] * program[program[index + 2]];
        } else {
          std::cout << "Error" << '\n';
        }
        index += 4;
      }
      if (program[0] == desired_output) {
        ans = noun * 100 + verb;
        std::cout << ans << '\n';
        return ans;
      }
    }
  }
  std::cout << ans << '\n';

  return ans;
}
